pages/index.tsx

```diff
@@ -0,0 +1,173 @@
+import Head from "next/head";
+import styles from "../styles/Home.module.css";
+import React, { useEffect, useRef, useState } from "react";
+import fetch from "node-fetch";
+import * as jwt from "jsonwebtoken";
+import {
+  CLIENT_ID,
+  HOSTED_UI_LOGIN_URL,
+  HOSTED_UI_TOKEN_URL,
+  REDIRECT_URI,
+} from "./api/constants";
+import {
+  DecodedIdToken,
+  IdDetails,
+  SsoDetails,
+  TokenResponse,
+} from "./api/types";
+import generateSaml from "./api/generateSaml";
+import CallCenterPicker from "../components/CallCenterPicker";
+import SelfSubmittingSsoForm from "../components/SelfSubmittingSsoForm";
+
+export default function Home() {
+  const [idDetails, setIdDetails] = useState<IdDetails | undefined>(undefined);
+  const [ssoDetails, setSsoDetails] = useState<SsoDetails | undefined>(
+    undefined,
+  );
+  const [pickerSubmitWasPressed, setPickerSubmitWasPressed] = useState(false);
+  const [singleSubmitTriggered, setSingleSubmitTriggered] = useState(false);
+  const [error, setError] = useState<string | undefined>(undefined);
+  // Control double effect-invocation when `fetch` should only happen once
+  const tokenFetchAlreadyRan = useRef<boolean>(false);
+
+  useEffect(() => {
+    // ensure `document` is going to be defined
+    if (typeof window !== "undefined") {
+      const authorizationCode = (
+        document.location.search
+          .substring(1)
+          .split("&")
+          .find((s) => s.match(/^code=/)) ?? ""
+      ).split("=")[1];
```

Comment:
this kind of string manipulation is always hard for me to follow... maybe a comment with an example original and result would make it clear?

Reply:
I shall make a standalone function!